packages/shared_preferences/shared_preferences_tool/lib/src/shared_preferences_tool_eval.dart

```diff
@@ -118,40 +138,142 @@ class SharedPreferencesToolEval {
     _valueDisposable?.dispose();
     _changeValueDisposable?.dispose();
     _removeValueDisposable?.dispose();
-    _eval.dispose();
+    _asyncEval.dispose();
+  }
+
+  Future<List<String>> _fetchAsyncKeys() async {
+    final Instance keysInstance = await _asyncEval.prefsGetInstance(
+      'getKeys()',
+      _isWeb,
+      _allKeysDisposable,
+    );
+    return Future.wait(<Future<String>>[
+      for (final InstanceRef keyRef
+          in keysInstance.elements!.cast<InstanceRef>())
+        _asyncEval.safeGetInstance(keyRef, _allKeysDisposable).then(
+              (Instance keyInstance) => keyInstance.valueAsString!,
+            ),
+    ]);
+  }
+
+  Future<List<String>> _fetchLegacyKeys() async {
+    final Instance keysSetInstance = await _legacyEval.legacyPrefsGetInstance(
+      'getKeys()',
+      _isWeb,
+      _allKeysDisposable,
+    );
+    return Future.wait(<Future<String>>[
+      for (final InstanceRef keyRef
+          in keysSetInstance.elements!.cast<InstanceRef>())
+        _legacyEval.safeGetInstance(keyRef, _allKeysDisposable).then(
+              (Instance keyInstance) => keyInstance.valueAsString!,
+            ),
+    ]);
+  }
+
+  Future<Instance> _getAsyncValue(String key) async {
+    return _asyncEval.prefsGetInstance(
+      "getAll(allowList: {'$key'}).then((map) => map.values.firstOrNull)",
+      _isWeb,
+      _valueDisposable,
+    );
+  }
+
+  Future<Instance> _getLegacyValue(String key) async {
+    return _legacyEval.legacyPrefsGetInstance(
+      "get('$key')",
+      _isWeb,
+      _valueDisposable,
+    );
   }
 }
 
 extension on EvalOnDartLibrary {
-  /// Evaluates the given [method] on the shared preferences instance.
+  Future<InstanceRef?> prefsEval(
+    String method,
+    bool isWeb,
+    Disposable? isAlive,
+  ) async {
+    return evalFuture(
+      'SharedPreferencesAsync().$method',
+      isWeb,
+      isAlive,
+    );
+  }
+
+  Future<Instance> prefsGetInstance(
+    String method,
+    bool isWeb,
+    Disposable? isAlive,
+  ) async {
+    return safeGetInstance(
+      (await prefsEval(method, isWeb, isAlive))!,
+      isAlive,
+    );
+  }
+
+  /// This only works on non-web platforms due to the `asyncEval` call.
+  /// This is ok, this won't ever be called on web platforms.
+  Future<InstanceRef?> legacyPrefsEval(
+    String method,
+    bool isWeb,
+    Disposable? isAlive,
+  ) {
+    return evalFuture(
+      'SharedPreferences.getInstance().then((prefs) => prefs.$method)',
```

Comment:
will we have any tests that verify that these evals works against version of shared_preferences the extension is shipped with?

Reply:
Hey @kenzieschmoll! That is a really good point. I remember that I tried the `fixture` test but I couldn't find the appropriate tooling exposed in `pub.dev`, but looking again I see that devtools_shared exposes the [cli_test_driver](https://github.com/flutter/devtools/blob/master/packages/devtools_shared/lib/devtools_test_utils.dart#L7) and [integration_test_runner](https://github.com/flutter/devtools/blob/master/packages/devtools_shared/lib/devtools_test_utils.dart#L7). Do you have any tips that could easy my research?

My initial idea was to create a `fixture` integration test using the [CliAppFixture](https://github.com/flutter/devtools/blob/181a95809fdab8f0584b660ea6788edd76406813/packages/devtools_shared/lib/src/test/cli_test_driver.dart#L78). But I see that it is only being used in legacy tests.

Could you share an example of a new integration test? Or is it ok to use `CliAppFixture`?

If we go with the `CliAppFixture` route I was planning on referencing the `shared_preferences` package as a relative path dependency in the fixture `pubspec.yaml`, then we'd always test against the same version that the extension is bundled to.

Reply:
This is difficult to test right now without forking some existing functionality from the DevTools codebase. I filed https://github.com/flutter/devtools/issues/8210 to track making this easier.

A really simple approach for now would be to add a test to `shared_preferences` that tests the expressions that are evaluated in `shared_preferences_tool`, and add some comments and code pointers to where these expressions are evaluated in the extension and warnings about modifying the test file or the classes / methods under test. While this isn't the most robust way to test this, this would at least cover the bases for now of avoiding accidental breakages that didn't even trigger a test failure with an explanation.

Reply:
🤔 If we just want to test that the "contract" works maybe we can do a test mocking the 'SharedPreferencesAsyncPlatform' and 'SharedPreferencesStorePlatform'. We could do something like this: When this expression is evaluated it should call this x method on `SharedPreferencesAsyncPlatform`.

I'll try it here and will let you know when I have something.

Reply:
My main concern is that someone working on `shared_preferences` may make a change that breaks one of the extension's expression evaluations, so we need to have some test coverage against that and that will run on each commit made to the `shared_preferences` package.

Reply:
> won't all packages re-run the tests on every change?

This is my understanding. @stuartmorgan can you confirm?

> if someone changed shared_preferences it would break the tool's test

Exactly, this is what we want.

Reply:
CI runs a package's tests against any PR that can affect that package (either because it touches the package, or because it touches something repo-wide).

Reply:
Does that cover this case since `shared_preferences_tool` is a subdirectory of `packages/shared_preferences`?

Reply:
No, it wouldn't because they are different packages. Sorry, I misunderstood the context of the question before.

[Your comment above](https://github.com/flutter/packages/pull/6749#discussion_r1722062005) seems like the best option for this case. The fact that it would require manual cross-referencing on failure shouldn't be an issue since it should only ever break when doing a major version bump, which is rare.